src/drawers/symbolDrawer.ts

```diff
@@ -4,15 +4,125 @@
  */
 
 import { Dataset } from "../core/dataset";
-
+import { Accessor, AttributeToAppliedProjector } from "../core/interfaces";
+import { SymbolFactory } from "../core/symbolFactories";
+import { CanvasBuffer } from "./canvasBuffer";
 import { Drawer } from "./drawer";
+import { AppliedDrawStep } from "./index";
 
 export class Symbol extends Drawer {
+  private buffer: CanvasBuffer;
 
-  constructor(dataset: Dataset) {
+  constructor(
+      dataset: Dataset,
+      private symbolProjector: () => Accessor<SymbolFactory>,
+      private sizeProjector: () => Accessor<number>,
+  ) {
     super(dataset);
     this._svgElementName = "path";
     this._className = "symbol";
   }
 
+  protected _drawStepCanvas(data: any[], step: AppliedDrawStep) {
+    this._drawSymbolsRubberStamp(data, step);
+  }
+
+  /**
+   * Uses an internal canvas buffer to blit the symbols into the canvas.
+   *
+   * For each symbol, we first detect if it's even in to the canvas bounds, if
+   * not we continue. Then we check it's other style attributes and symbol type.
+   * If those don't match what are in the buffer, we redraw the buffer. Finally,
+   * we blit the buffer to the canvas at the symbol coordinates.
+   */
+  private _drawSymbolsRubberStamp(data: any[], step: AppliedDrawStep) {
+    const ctx = this.canvas().node().getContext("2d");
+    const projector = step.attrToAppliedProjector;
+    const styleKeys = Object.keys(Drawer._CANVAS_CONTEXT_ATTRIBUTES);
+    const attrKeys = styleKeys.concat(["x", "y"]);
+
+    // create canvas intersection tester
+    const { width, height } = ctx.canvas; // TODO devicePixelRatio?
+    const intersectsCanvasBounds = (x: number, y: number, size: number) => {
+      return (
+        x + size >= 0 && x - size <= width &&
+        y + size >= 0 && y - size <= height
+      );
+    };
+
+    // lazilly create offscreen buffer of modest size
+    if (this.buffer == null) {
+      this.buffer = new CanvasBuffer(10, 10, 1);
+      this.buffer.ctx.translate(5, 5);
+    }
+
+    let prevAttrs: any = null;
+    let prevSymbolGenerator: any = null;
+    let prevSymbolSize: number = null;
+    let skipped = 0;
+    for(let index = 0; index < data.length; index++) {
+      const datum = data[index];
+
+      // check symbol is in viewport
+      const attrs = this._resolveAttributesSubset(projector, attrKeys, datum, index);
+      const symbolSize = this.sizeProjector()(datum, index, this._dataset);
+      if (!intersectsCanvasBounds(attrs["x"], attrs["y"], symbolSize * 2)) {
```

Comment:
This point is set to `(0,0)` in the data

![screen shot 2017-03-22 at 13 40 27](https://cloud.githubusercontent.com/assets/6375593/24200512/2beecb04-0f05-11e7-8363-1427ef58a6b8.png)